src/libopensc/pkcs15.c

```diff
@@ -1156,6 +1190,20 @@ sc_pkcs15_bind_internal(struct sc_pkcs15_card *p15card, struct sc_aid *aid)
 
 	*(p15card->tokeninfo) = tokeninfo;
 
+	/* If card has not created algorithms, let it look at token info */
+	sc_log(ctx, " card->algorithms:%p card->algorithm_count:%d", card->algorithms, card->algorithm_count);
+	if (!card->algorithms && card->algorithm_count == 0) {
+		struct sc_cardctl_parsed_token_info  parsed_token_info;
+
+		sc_log(ctx, " passing sc_cardctl_parsed_token_info ");
+		/* TODO could set flags here */
+		parsed_token_info.flags = 0;
+		parsed_token_info.tokeninfo = p15card->tokeninfo;
+		parsed_token_info.supported_algos = &p15card->tokeninfo->supported_algos;
+		sc_card_ctl(p15card->card, SC_CARDCTL_PARSED_TOKEN_INFO, &parsed_token_info);
```

Comment:
Did you look at `pkcs15_prkey_can_do()`? It looks like PKCS#11 is already checking the private keys' `struct sc_pkcs15_prkey_info.algo_refs` against `tokeninfo->supported_algos`. Since this is currently only done when the private key refers to a specific algoritm from TokenInfo, you could change this in a generic manner by first checking if the algorithms appears in TokenInfo at all. The change should look something like this:
```diff
diff --git a/src/pkcs11/framework-pkcs15.c b/src/pkcs11/framework-pkcs15.c
index dae65c55f..12ea0611e 100644
--- a/src/pkcs11/framework-pkcs15.c
+++ b/src/pkcs11/framework-pkcs15.c
@@ -4308,16 +4308,27 @@ pkcs15_prkey_can_do(struct sc_pkcs11_session *session, void *obj,
 	if (!prkey || !prkey->prv_info)
 		return CKR_KEY_FUNCTION_NOT_PERMITTED;
 
-	pkinfo = prkey->prv_info;
-	/* Return in there are no usage algorithms specified for this key. */
-	if (!pkinfo->algo_refs[0])
-		return CKR_FUNCTION_NOT_SUPPORTED;
-
 	if (!p11card)
 		return CKR_TOKEN_NOT_RECOGNIZED;
 	fw_data = (struct pkcs15_fw_data *) p11card->fws_data[session->slot->fw_data_idx];
 	token_algos = &fw_data->p15_card->tokeninfo->supported_algos[0];
 
+	if (token_algos->reference) {
+		/* The profile announces some capabilities.
+		 * Check if the algorithm is supported by the card at all */
+		for (jj=0;jj<SC_MAX_SUPPORTED_ALGORITHMS && (token_algos + jj)->reference; jj++) {
+			if ((token_algos + jj)->mechanism == mech_type)
+				break;
+		}
+		if (jj == SC_MAX_SUPPORTED_ALGORITHMS || !(token_algos + jj)->reference)
+			return CKR_MECHANISM_INVALID;
+	}
+
+	pkinfo = prkey->prv_info;
+	/* Return in there are no usage algorithms specified for this key. */
+	if (!pkinfo->algo_refs[0])
+		return CKR_FUNCTION_NOT_SUPPORTED;
+
 	for (ii=0;ii<SC_MAX_SUPPORTED_ALGORITHMS && pkinfo->algo_refs[ii];ii++)   {
 		/* Look for algorithm supported by token referenced in the list of key's algorithms */
 		for (jj=0;jj<SC_MAX_SUPPORTED_ALGORITHMS && (token_algos + jj)->reference; jj++)
```

Reply:
... this would then solve this issue for all other cards as well and it doesn't introduce a new callback.

Reply:
I have not looked at  `pkcs15_prkey_can_do()` 

I have been looking only at cardos cards. Some like @Jakuje s card have no supporteded_algos, One of the other says it has PKCS1, most says RAW RSA. one has 8. 

 OpenSC looks for "keyReference" and "algReference" in `pkcs15-prkey.c`:
```
80         { "keyReference",SC_ASN1_INTEGER, SC_ASN1_TAG_INTEGER, SC_ASN1_OPTIONAL, NULL, NULL },
 81 /* Absent in PKCS#15-v1.1 but present in ISO 7816-15(2004-01-15)*/
 82         { "algReference", SC_ASN1_STRUCT, SC_ASN1_CONS | SC_ASN1_CTX | 1, SC_ASN1_OPTIONAL, NULL, NULL },
``` 
No cardos card appears to have a algReference.

Reply:
The change that I've proposed checks the global algorithms in TokenInfo, which is meant to replace your explicit call to
```
 		sc_card_ctl(p15card->card, SC_CARDCTL_PARSED_TOKEN_INFO, &parsed_token_info);
```
You've implemented cardos_cardctl_parsed_token_info() to add algorithms only when they appear in TokenInfo.

My suggestion announces all algorithms as before, but when the PKCS#11 initializes a signature with a specific key (`C_SignatureInit`), OpenSC calls `key->ops->can_do` (`pkcs15_prkey_can_do()`), which then checks TokenInfo for the global mechanisms (and the key specific algorithms as before). Effectively, the application will receive an error if the card doesn't support that algorithm in TokenInfo **or** if the key is restricted to not support this algorithm.

Reply:
Remember that I do not have a Cardos card and am relying on those who have cards. Some of the cards appear to only support PKCS1 on the card, others RAW RSA. TokenInfo helps. 
But some cards don't have a full TokenInfo with no supported algos, and just the EIDComplient bit set.  

Some on this list accuse the card's TokenInfo of "lying".  The "lying" may be in how the vendor or the provisioner of the card interpreted PKCS15 and wrote just the minimal (or incorrect) information and use other non-PKCS15 nformation to fill in the mechanism supported. (We don't have any documentation on these cards.)

So I don't see a general solution without getting the card driver involved. 

You may be right that your code will work, but my goal is to get a working solution for CardOS cards people have today.  When I am done, feel free to add your changes and rip out the SC_CARDCTL_PARSED_TOKEN_INFO and related code.

Reply:
The built-in way of doing this would be to have a PKCS#15 emulator for CardOS bootstrapped from the synthetic way of reading the card's object. This allows parsing/modifying all such objects.

However, I still think we are missing some information from the vendor documentation here. I doubt that ATOS' middleware really has exceptions for potentially thousands of card profiles. I still believe there's a generic way that doesn't involve EIDComplient...